rustls/src/conn.rs

```diff
@@ -1316,6 +1318,35 @@ impl CommonState {
         #[cfg(not(feature = "quic"))]
         false
     }
+
+    pub(crate) fn should_update_key(
+        &mut self,
+        key_update_request: &KeyUpdateRequest,
+    ) -> Result<bool, Error> {
+        match key_update_request {
+            KeyUpdateRequest::UpdateNotRequested => Ok(false),
+            KeyUpdateRequest::UpdateRequested => Ok(self.queued_key_update_message.is_none()),
+            _ => {
+                self.send_fatal_alert(AlertDescription::IllegalParameter);
+                Err(Error::CorruptMessagePayload(ContentType::Handshake))
+            }
+        }
+    }
+
+    pub(crate) fn enqueue_key_update_notification(&mut self) {
+        let message = PlainMessage::from(Message::build_key_update_notify());
+        self.queued_key_update_message = Some(
```

Comment:
Why do we have the indirection through `self.queued_key_update_message`? Why not `append()` to `self.sendable_tls` directly here?

Reply:
```
   If the request_update field is set to "update_requested", then the
   receiver MUST send a KeyUpdate of its own with request_update set to
   "update_not_requested" prior to sending its next Application Data
   record.  This mechanism allows either side to force an update to the
   entire connection, but causes an implementation which receives
   multiple KeyUpdates while it is silent to respond with a single
   update.  Note that implementations may receive an arbitrary number of
   messages between sending a KeyUpdate with request_update set to
   "update_requested" and receiving the peer's KeyUpdate, because those
   messages may already be in flight.  However, because send and receive
   keys are derived from independent traffic secrets, retaining the
   receive traffic secret does not threaten the forward secrecy of data
   sent before the sender changed keys.
```

I think it's the "causes an implementation which receives multiple KeyUpdates while it is silent to respond with a single update" clause that we're trying to do here.

Reply:
Ahh, so we only want to respond to `KeyUpdateRequest`s once and only send the `KeyUpdate` if we have something else to send, too. So would be good to add some comments to clarify this.

Reply:
I think that's the intent of `// Mustn't be interleaved with other handshake messages.`, which was present in one call site but not the other, and I'll duplicate. There is, however, a bogo test that fails if we just directly enqueue the message without keeping track of that state

Reply:
Added a comment, would appreciate feedback on accuracy and clarity

Reply:
I think deduplication is one thing but I would probably also mention that we're trying to avoiding sending a key update notification if there isn't anything else to send.